### ISSUE TITLE
Static Type Checking

### ISSUE THREAD

**Issue by MeirionHughes** [MeirionHughes/aurelia-template-lint#24]
Proposal:
* introduce static type syntax to bindables in template html. 
* Type syntax should be stripped out of output html. 

It would be nice to be able to check that bindings are correct, at build-time. For example. Given the following two files. 
 
*foo.html*
```html
<template bindable="person:Person">
   ${person.nme}  <!-- oops -->
</template>
```
*person.ts*
```ts
export class Person{
   name:string
}
```
Then, it should be detectable that `person.nme` is invalid. 

Building a system that can achieve this, without actually fully running aurelia should be achievable. Possible solution was discussed here: https://github.com/MeirionHughes/aurelia-template-lint/issues/4#issuecomment-225514323 .

**Comment by MeirionHughes** [comment]
* [ ] investigate [using the typescript compiler to traverse the syntax tree](https://github.com/Microsoft/TypeScript/wiki/Using-the-Compiler-API-%28TypeScript-1.4%29)

**Comment by atsu85** [comment]
You probably want much higher level interface to deal with TypeScript classes (to fetch the fields and methods and their types). I guess TypeScript already provides that information for IDE's to create an outline of the class, that includes its method and fields - maybe it is implemented somewhere [here](https://github.com/Microsoft/TypeScript/tree/master/src/server)?

**Comment by MeirionHughes** [comment]
Had a quick look. Seems higher level with regard to the project overall, but not seeing reflection style information there, unless its imported from CompilerServices or LanguageServices?

I quite like the look of the AST node. Pretty much hands you the entire reflection data for files. should be super easy to pull out class/interface names and their variables. 

As you say, I need classes/interfaces and a list of their fields, getters and constructor created properties. So far that AST stuff is all I know of that does that.

**Comment by atsu85** [comment]
> So far that AST stuff is all I know of that does that. 

Yeah, I was just hoping, that You might find a tool (built on top of low-level "AST stuff") that already provides You the same high-level information.You'd need. It just seems unbelievable, that all the IDE plugins would implement the same thing for "Class outline view" without relying on higher level tools provided by TypeScript. 

> but not seeing reflection style information there, unless its imported from CompilerServices or LanguageServices?

Maybe, i guess it has to be somewhere  ;)

**Comment by MeirionHughes** [comment]
right, the reflection is there: via client `getProgram` which contains a list of `SourceFile`, which gives you: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts#L1601

**Comment by MeirionHughes** [comment]
this is actually super easy; 

```ts
export class Person{
    forename:string;    
}
```
`npm install typescript`

`ts = require('typescript')`
`fs = require('fs').readFileSync`
`file = ts.createSourceFile("test.ts", fs("test.ts").toString(), ts.ScriptTarget.ES7, true)`
`file.externalModuleIndicator.members[0].name.text`

gets you `forename`; Reflection is all readily available.

**Comment by MeirionHughes** [comment]
HTML reflection: https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L88

**Comment by zakjan** [comment]
I see for this milestone you are aiming to support view-only templates. This is a great start, looks really promising! I already integrated the linter to my quite data-intensive project, I'll test every new release :)

The next goal should be to support view-models as well:

```
export class PageViewModel {
    @bindable person: Person;
    title: string;
}

<template>
    ${titl}
    ${person.nme}
    ${person.name.toUpperCase()}
</template>
````

And in the end all Aurelia valid bindings.

**Comment by MeirionHughes** [comment]
Yeah, that would be the next logical step; I don't see it being a big issue if you have a 1-1 file relationship: `person.html` and `person.ts` and say it exports `PersonViewModel` That would be fairly straight forward.

**Comment by MeirionHughes** [comment]
@zakjan make sure you lock down the versions with anything production. aurelia-template-lint locks down on minor version (updates patches) of `template-lint`, so in principle when I break `template-lint` on next release it should be fine. 

I currently break on minor releases, while no breaking changes on minor.

**Comment by MeirionHughes** [comment]
* blocked by: https://github.com/aurelia/pal-nodejs/issues/1

**Comment by JeroenVinke** [comment]
@MeirionHughes made a proof of concept [here](https://github.com/JeroenVinke/aurelia-template-lint-typescript). I'd love to get Aurelia to give us more information about a view (a total list of bindings) so we don't have to traverse the DOM and fetch the binding expressions ourselves. Perhaps that will require browser capabilities though

**Comment by JeroenVinke** [comment]
Looks like you're miles ahead of me :smile:

**Comment by MeirionHughes** [comment]
Actually, I think your approach is better; We're already traversing the html tree in this linter anyway. My issue was trying to find a way grab the reflection of bindings, via the binding system. Looks like you've actually done that in your proof-of-concept. I find that extremely useful, Thank you!

I may "borrow" your approach in this regard and get a working Rule up and running.

**Comment by JeroenVinke** [comment]
If possible, i'd rather let aurelia parse an entire view and give us back the controller (I think a controller contains all the bindings etc). That would give us a lot more information than we could get by traversing the DOM. I think we need this info (if not now, then later) because aurelia views can be complex (if.bind's, <composes>, custom elements, attributes etc).

**Comment by MeirionHughes** [comment]
Yeah, it would ultimately be better to use the ViewCompiler in the long run. In that regard there I just need to implement the Mutation Observer emulation for the nodejs platform, test it and it will be good to go.

**Comment by MeirionHughes** [comment]
@JeroenVinke finally managed to at least boot up aurelia's template binding and extract the attribute binding information.

**Comment by JeroenVinke** [comment]
Awesome!! Good work!

**Comment by MeirionHughes** [comment]
![giphy](https://cloud.githubusercontent.com/assets/3584509/16226832/2dd00d30-37a5-11e6-9cd0-417e92236ca2.gif)

... to be fair, I'm pretty sure there are quite a few issues to iron out. :P
